### ISSUE TITLE
GIF image transcoding speed is too slow and quality is too large

### ISSUE THREAD

**Issue by jackDiscovery** [gographics/imagick#220]
The gographics version is 6.9.1-7
 The imageMagick version is 7.0.7.11

When I use imagemagick transcoding, I found that the quality of the gif image after transcoding is very high and the speed is very slow. The transcoding of a 50-frame image takes 3s. After testing, it is found that the following function is relatively slow:   
    `retImg = imgWand.GetImagesBlob()`

I saw in the community that you can optimize in the following way, the command is as follows:
```
convert xxx.gif $completeMiff -dither none -deconstruct \
    -layers optimize -matte -depth 7 -resize 100x73 xxx.gif
```
The reference is：http://studio.imagemagick.org/discourse-server/viewtopic.php?f=1&t=13339&start=0
But I did not find an optimized method in the Go code. I combined it according to the API, and the code is as follows:

```go
imgWand = imgWand.MergeImageLayers(imagick.IMAGE_LAYER_OPTIMIZE)
err := imgWand.PosterizeImage(uint(2), imagick.DITHER_METHOD_NO)
if err != nil {
    return nil, false, NewError(exception.BadImageRequestBody, "Image PosterizeImage")
}
err = imgWand.SetImageMatte(true)
if err != nil {
    return nil, false, NewError(exception.BadImageRequestBody, "Image SetImageMatte")
}
err = imgWand.SetImageDepth(uint(7))
if err != nil {
    return nil, false, NewError(exception.BadImageRequestBody, "Image SetImageDepth")
}                
```

![20191217-102931](https://user-images.githubusercontent.com/33868757/70962344-c7bb3a80-20bf-11ea-9c29-0d9a04d5c262.gif)

**Comment by justinfx** [comment]
> The gographics version is 6.9.1-7
The imageMagick version is 7.0.7.11

Can you confirm if this means you are using the ImageMagick7 `convert` command, versus the Go imagick bindings against ImageMagick 6? Have you tried comparing it against Go bindings using ImageMagick-7?

> But I did not find an optimized method in the Go code

Is this the method you are looking for?
https://godoc.org/gopkg.in/gographics/imagick.v3/imagick#MagickWand.OptimizeImageLayers

**Comment by jackDiscovery** [comment]
```
The gographics version is 6.9.1-7
The imageMagick version is 7.0.7.11
```
it means using Go bindings using ImageMagick-7

**Comment by jackDiscovery** [comment]
```
convert xxx.gif $completeMiff -dither none -deconstruct \
    -layers optimize -matte -depth 7 -resize 100x73 xxx.gif
```
What does the $ completeMiff parameter mean?
  Is there a corresponding method in Go imagick?

**Comment by justinfx** [comment]
> ```
> The gographics version is 6.9.1-7
> The imageMagick version is 7.0.7.11
> ```
> 
> it means using Go bindings using ImageMagick-7

It still doesn't make sense to me. Either you are using gographics/imagick.v2 with ImageMagick6 or you are using gographics/imagick.v3 with ImageMagick7. I'm not clear what the 6.x version means here.

**Comment by justinfx** [comment]
`$completeMiff` is called a shell variable. The person that wrote that command example was substituting real file paths for variables. They are not related to ImageMagick functionality here.

**Comment by jackDiscovery** [comment]
The current problem is that the gif image is scaled and transcoded, which took more than 2 seconds as a whole.
`retImg = imgWand.GetImagesBlob()`

Calling this method took 2s
I don't know if there is a solution to improve the speed of transcoding gifs?

**Comment by justinfx** [comment]
I did a test using a large 500x375 323 frame source gif, and I got the same timings for the Go bindings code vs the equivalent `convert` command.

Source image:
https://lh5.googleusercontent.com/-mvpIxrtlkAA/Tndmi1sf6zI/AAAAAAAAD1E/I2vwpRqn1z8/s500/FlyEarth500_Micael_Reynaud.gif

```go
package main

import (
	"fmt"
	"time"

	"gopkg.in/gographics/imagick.v3/imagick"
)

func main() {
	imagick.Initialize()
	defer imagick.Terminate()
	var err error

	mw := imagick.NewMagickWand()

	start := time.Now()
	slice := start

	err = mw.ReadImage("source.gif")
	if err != nil {
		panic(err)
	}
	fmt.Printf("Read: %v\n", time.Since(slice))

	err = mw.PosterizeImage(uint(2), imagick.DITHER_METHOD_NO)
	if err != nil {
		panic(err)
	}

	slice = time.Now()
	mw = mw.OptimizeImageLayers()
	fmt.Printf("OptimizeImageLayers: %v\n", time.Since(slice))

	err = mw.SetImageMatte(true)
	if err != nil {
		panic(err)
	}
	err = mw.SetImageDepth(uint(7))
	if err != nil {
		panic(err)
	}

	hWidth := uint(mw.GetImageWidth() / 4)
	hHeight := uint(mw.GetImageHeight() / 4)

	mw.ResetIterator()
	for {
		err = mw.ResizeImage(hWidth, hHeight, imagick.FILTER_LANCZOS)
		if err != nil {
			panic(err)
		}
		if !mw.NextImage() {
			break
		}
	}

	err = mw.WriteImages("out.gif", true)
	if err != nil {
		panic(err)
	}

	fmt.Printf("Total: %v\n", time.Since(start))
}
```

Results:
```
$ go run main.go
Read: 1.880297815s
OptimizeImageLayers: 6.173776442s
Total: 49.62787901s

$ time convert source.gif -dither none -deconstruct -layers optimize -matte -depth 7 -resize 125x93 out2.gif 
...  142.53s user 17.16s system 323% cpu 49.317 total
```

There wasn't much of a difference between calling `WriteImages()` to a file, vs `GetImagesBlob()` to get the bytes in memory.

**Comment by jackDiscovery** [comment]
I found that this command can improve the processing speed of gif
```
time convert pig-bug.gif -coalesce  -filter box  -resize 100x73  -layers Optimize pig_thumbnail_3.gif

real	0m0.428s
user	0m0.423s
sys	0m0.077s
```
But using Go bindings is still slow.
```
             imgWand := imagick.NewMagickWand()
		 err = imgWand.ReadImageBlob(imgProcessor.SourceImage)
		 if err != nil {
			return nil
	         }
		 imageNum := imgWand.GetNumberImages()
		 imgWand = imgWand.CoalesceImages()
		 bordColor := imagick.NewPixelWand()
		 defer bordColor.Destroy()
		 bordColor.SetColor("#b0c4de")
		 //bordColor.SetColor("LightSteelBlue")
		 err := imgWand.SetImageBorderColor(bordColor)
		 if err != nil {
		 	return nil
		 }
                for index := 0; index < int(imageNum); index++ {
				imgWand.SetIteratorIndex(index)
				watermark, ok := operation.(WatermarkOperation)
				imgErr := operation.process(imgWand)
				if imgErr != nil {
					return nil, false, imgErr
				}
		}
		imgWand = imgWand.OptimizeImageLayers()
		startCodeTime2 := time.Now()
		retImg := imgWand.GetImagesBlob()
		errorLogger.Printf("GetImageBlobs used:%.0fms.", 
                                time.Now().Sub(startCodeTime2).Seconds()*1000)
```

`GetImageBlobs used:1980ms.`

I don't know why？

**Comment by justinfx** [comment]
I don't know why either. How long does it take when you write the image instead of getting the blob? How long does it take when you do the watermarking with the convert command?
I'm not sure what ImageMagick does when it gets the final image after you apply operations to the images.

**Comment by jackDiscovery** [comment]
```
time convert pig-bug.gif -coalesce  -filter box  -resize 100x73  -layers Optimize pig_thumbnail_3.gif

real	0m0.431s
user	0m0.428s
sys	0m0.074s
```

```
time convert pig-bug.gif -coalesce  -filter box -layers Optimize -resize 100x73  pig_thumbnail_3.gif

real	0m3.081s
user	0m2.280s
sys	0m1.241s
```
Put the -layers Optimize command before and after the -resize command, the processing time is very different

**Comment by justinfx** [comment]
Makes sense to me.

**Comment by jackDiscovery** [comment]
convert pig-bug.gif -coalesce -filter box -resize 100x73 -layers Optimize pig_thumbnail_3.gif

How to write the above command in Golang language, please give me a demo

**Comment by justinfx** [comment]
Well I would have thought to write it like this:

```go
package main

import (
	"fmt"
	"time"

	"gopkg.in/gographics/imagick.v3/imagick"
)

func main() {
	imagick.Initialize()
	defer imagick.Terminate()
	var err error

	mw := imagick.NewMagickWand()

	start := time.Now()
	slice := start

	err = mw.ReadImage("source.gif")
	if err != nil {
		panic(err)
	}
	fmt.Printf("Read: %v\n", time.Since(slice))

	mw = mw.CoalesceImages()

	hWidth := uint(mw.GetImageWidth() / 4)
	hHeight := uint(mw.GetImageHeight() / 4)

	slice = time.Now()
	mw.ResetIterator()
	for {
		err = mw.ResizeImage(hWidth, hHeight, imagick.FILTER_BOX)
		if err != nil {
			panic(err)
		}
		if !mw.NextImage() {
			break
		}
	}
	fmt.Printf("ResizeImages: %v\n", time.Since(slice))

	slice = time.Now()
	mw = mw.OptimizeImageLayers()
	fmt.Printf("OptimizeImageLayers: %v\n", time.Since(slice))

	slice = time.Now()
	err = mw.WriteImages("out.gif", true)
	if err != nil {
		panic(err)
	}
	fmt.Printf("Write: %v\n", time.Since(slice))

	fmt.Printf("Total: %v\n", time.Since(start))
}
```

But I can see that it spends a very long time on the `WriteImages`:

```
$ go run main.go
Read: 1.070154816s
ResizeImages: 719.499337ms
OptimizeImageLayers: 127.098351ms
Write: 19.64175276s
Total: 22.316453501s
```

If I set `export MAGICK_DEBUG=All` in my environment, I can see that when I run the Go application it is doing a lot more work during the Coder step of writing out the gif images, compared to the `convert` command seeming to do more caching and less work. I am not sure what the difference is.

**Comment by jackDiscovery** [comment]
I have found a way to solve it, C provides a method C.ConvertImageCommand
package imagick

Below is my code using cgo

```
/*
#include <unistd.h>
#include <MagickWand/MagickWand.h>
#include <stdio.h>
*/
import "C"
 
import (
	"unsafe"
)
 
 
// ConvertImageCommand 命令行转码
func  ConvertImageCommand(args []string) error {
	size := len(args)
	arg := make([]*C.char, size)
	for i,_ := range args{
		arg[i] = C.CString(args[i])
		defer C.free(unsafe.Pointer(arg[i]))
	}
	argStr := make([]*C.char, size)
	for i,_ := range args{
		argStr[i] = C.CString(args[i])
		defer C.free(unsafe.Pointer(argStr[i]))
	}
	imageinfo := newImageInfo()
	defer C.DestroyImageInfo(imageinfo.info);
	var exc *C.ExceptionInfo = C.AcquireExceptionInfo();
	defer C.DestroyExceptionInfo(exc);
	ok := C.ConvertImageCommand(imageinfo.info, C.int(size),(**C.char)(unsafe.Pointer(&arg[0])),(**C.char)(unsafe.Pointer(&argStr[0])),exc)
	if C.int(ok) == 0 {
		exception := newExceptionInfo(exc)
		return  exception
	}
	return nil
}
```
The problem is solved, I still don't understand, wait for time to see the code of c.
It is recommended to add this method to subsequent versions

**Comment by justinfx** [comment]
You could ask on the ImageMagick forum to see if anyone knows what the C API is doing in that command, compared to your API example that is slow.

Yes we should probably expose `ConvertImageCommand`. Good find.

**Comment by justinfx** [comment]
I'm working on adding `ConvertImageCommand`

**Comment by justinfx** [comment]
You can give this feature a try in the [220_convert_command](https://github.com/gographics/imagick/tree/220_convert_command) branch (im7 only). I can merge it if you are happy with the way it is exposed in the API. I could look into adding it for im6 as well, but it requires us adding the missing `ExceptionInfo` type as well to that branch

**Comment by justinfx** [comment]
Merged to im7 branch

**Comment by justinfx** [comment]
Released as v2.6.0 and v3.3.0
